src/formatter/simple.rs

```diff
@@ -1,361 +1,433 @@
+use std::io;
+use std::mem;
+
+use colored::*;
+use context::{ContextInfo, ExampleInfo};
 use events::{Event, EventHandler};
 use formatter::formatter::Formatter;
-use runner;
-use std::io;
+use report::suite::SuiteReport;
+use report::context::ContextReport;
+use report::example::ExampleReport;
+use suite::SuiteInfo;
+
+#[derive(Clone, Debug)]
+enum ScopeInfo {
+    Suite(SuiteInfo),
+    Context(ContextInfo),
+    Example(ExampleInfo),
+}
 
-pub struct Simple<'a, Io: io::Write + 'a> {
-    buf: &'a mut Io,
-    pub name_stack: Vec<String>,
-    pub failures: Vec<String>,
+pub struct Simple<T: io::Write> {
+    buffer: T,
+    name_stack: Vec<ScopeInfo>,
+    failed: Vec<Vec<ScopeInfo>>,
 }
 
-impl<'a, T: io::Write> Simple<'a, T> {
-    pub fn new(buf: &mut T) -> Simple<T> {
+impl<T: io::Write> Simple<T> {
+    pub fn new(buffer: T) -> Simple<T> {
         Simple {
-            buf: buf,
+            buffer: buffer,
             name_stack: vec![],
-            failures: vec![],
+            failed: vec![],
         }
     }
 
-    fn failures_summary(&self) -> String {
-        let res = String::with_capacity(100);
-        let mut idx = 0;
-        self.failures
-            .iter()
-            .map(|fail| {
-                idx += 1;
-                format!("  {}) {}\n", idx, fail)
-            })
-            .fold(res, |mut acc, elt| {
-                acc.push_str(&elt);
-                acc
-            })
+    fn padding(depth: usize) -> String {
+        "  ".repeat(depth)
     }
 
-    fn write_summary(&mut self, result: runner::RunnerResult) -> String {
-        let (res, report) = match result {
-            Ok(report) => ("ok", report),
-            Err(report) => ("FAILED", report),
-        };
-
-        format!("\n\ntest result: {}. {} examples; {} passed; {} failed;",
-                res,
-                report.total_tests,
-                report.success_count,
-                report.error_count)
+    fn enter_suite(&mut self, info: &SuiteInfo) {
+        self.name_stack.push(ScopeInfo::Suite(info.clone()));
+        let _ = writeln!(self.buffer, "\nrunning tests");
+        let label: &str = info.label.into();
+        let _ = writeln!(self.buffer, "{} {:?}:", label, info.name);
     }
-}
 
-impl<'a, T: io::Write> EventHandler for Simple<'a, T> {
-    fn trigger(&mut self, event: &Event) {
-        // FIXME: do something with the io::Error ?
-        let _ = match *event {
-            Event::StartRunner => writeln!(self.buf, "Running tests:\n"),
-            Event::StartDescribe(ref name) |
-            Event::StartTest(ref name) => {
-                self.name_stack.push(name.clone());
-                Ok(())
-            }
-            Event::EndTest(result) => {
-                if result.is_err() && !self.name_stack.is_empty() {
-                    let failure_name = self.name_stack.join(" | ");
-                    self.failures.push(failure_name);
+    fn exit_suite(&mut self, report: &SuiteReport) {
```

Comment:
Yes. The way `formatter::Simple` (which by now should probably be renamed to `formatter::Standard`, as it's not simple any more) is implemented right now isn't really compatible with parallel execution anyway.

The `name_stack: Vec<ScopeInfo>` should not be managed inside the formatter, but the `Runner` should manage it along with the environments being passed around and then pass the name stack to the formatter by reference via events. I'm on it.